File: app/src/main/java/com/moisesborges/tvaddict/showdetails/ShowDetailsPresenter.java
package com.moisesborges.tvaddict.showdetails;

import android.support.annotation.NonNull;

import com.moisesborges.tvaddict.data.ShowsRepository;
import com.moisesborges.tvaddict.models.Season;
import com.moisesborges.tvaddict.models.Show;
import com.moisesborges.tvaddict.mvp.BasePresenter;
import com.moisesborges.tvaddict.mvp.RxJavaConfig;

import javax.inject.Inject;

import io.reactivex.disposables.Disposable;

/**
 * Created by moises.anjos on 18/04/2017.
 */

public class ShowDetailsPresenter extends BasePresenter<ShowDetailsView> {

    private final ShowsRepository mShowsRepository;

    @Inject
    public ShowDetailsPresenter(@NonNull ShowsRepository showsRepository,
                                @NonNull RxJavaConfig rxJavaConfig) {
        super(rxJavaConfig);
        mShowsRepository = showsRepository;
    }

    public void loadShowDetails(@NonNull Show show) {
        checkView();

        displayBasicShowInfo(show);

        loadShowAdditionalInfo(show);

        verifySavedShowStatus(show);
    }

    private void displayBasicShowInfo(@NonNull Show show) {
        getView().setShowImage(show.getImage().getMedium());
        getView().setShowName(show.getName());
        getView().setShowSummary(show.getSummary());
        getView().displayAdditionalInfoNotLoaded(false);
        getView().setShowRating(show.getRating() != null ? show.getRating().getAverage() : null);
        getView().setShowNetwork(show.getNetwork() != null ? show.getNetwork().getName() : null);
        getView().setShowRuntime(show.getRuntime());
        getView().setShowGenres(show.getGenres());
        getView().setShowExternalLinks(show.getExternals());
    }

    private void loadShowAdditionalInfo(@NonNull Show show) {
        getView().displayAdditionalInfoLoadingInProgress(true);
        Disposable loadShowDataDisposable = mShowsRepository.getFullShowInfo(show.getId())
                .compose(applySchedulersToSingle())
                .onErrorReturn(ignored -> Show.NOT_FOUND)
                .subscribe(showFullInfo -> {
                    getView().displayAdditionalInfoLoadingInProgress(false);
                    if (showFullInfo == Show.NOT_FOUND) {
                        getView().displayAdditionalInfoNotLoaded(true);
                        return;
                    }

                    getView().setShow(showFullInfo);
                    if (showFullInfo.getSeasons() != null) {
                        getView().displaySeasons(showFullInfo.getSeasons());
                    }
                    if (showFullInfo.getCast() != null) {
                        getView().displayCastMembers(showFullInfo.getCast());
                    }
                });

        addDisposable(loadShowDataDisposable);
    }

    private void verifySavedShowStatus(@NonNull Show show) {
        Disposable checkSavedShowDisposable = mShowsRepository.getSavedShow(show.getId())
                .compose(applySchedulersToSingle())
                .subscribe(savedShow -> getView().displaySaveShowButton(false),
                        ignored -> getView().displaySaveShowButton(true));

        addDisposable(checkSavedShowDisposable);
    }

    public void changeWatchingStatus(@NonNull Show show) {
        checkView();

        Disposable checkStatusDisposable = mShowsRepository.getSavedShow(show.getId())
                .compose(applySchedulersToSingle())
                .onErrorReturn(ignored -> Show.NOT_FOUND)
                .flatMap(showFromDb -> {
                    if (showFromDb == Show.NOT_FOUND) {
                        return mShowsRepository.saveShow(show);
                    } else {
                        return mShowsRepository.removeShow(show.getId());
                    }
                })
                .subscribe(showAfterChange -> {
                    if (showAfterChange == Show.REMOVED) {
                        getView().displayShowRemovedMessage();
                        getView().displaySaveShowButton(true);
                    } else {
                        getView().displaySavedShowMessage();
                        getView().displaySaveShowButton(false);
                    }

                });

        addDisposable(checkStatusDisposable);
    }

    public void openEpisodes(Show show, Season season) {
        checkView();

        getView().navigateToEpisodes(show, season.getNumber());
    }
}

File: app/src/main/java/com/moisesborges/tvaddict/showdetails/ShowDetailsPresenter.kt
package com.moisesborges.tvaddict.showdetails

import com.moisesborges.tvaddict.data.ShowsRepository
import com.moisesborges.tvaddict.models.Season
import com.moisesborges.tvaddict.models.Show
import com.moisesborges.tvaddict.mvp.BasePresenter
import com.moisesborges.tvaddict.mvp.RxJavaConfig

import javax.inject.Inject

/**
 * Created by moises.anjos on 18/04/2017.
 */

class ShowDetailsPresenter @Inject
constructor(private val showsRepository: ShowsRepository,
            rxJavaConfig: RxJavaConfig) : BasePresenter<ShowDetailsView>(rxJavaConfig) {

    fun loadShowDetails(show: Show) {
        checkView()

        displayBasicShowInfo(show)

        loadShowAdditionalInfo(show)

        verifySavedShowStatus(show)
    }

    private fun displayBasicShowInfo(show: Show) {
        view.setShowImage(show.image.medium)
        view.setShowName(show.name)
        view.setShowSummary(show.summary)
        view.displayAdditionalInfoNotLoaded(false)
        view.setShowRating( show.rating?.average)
        view.setShowNetwork( show.network?.name ?: "")
        view.setShowRuntime(show.runtime)
        view.setShowGenres(show.genres)
        view.setShowExternalLinks(show.externals)
    }

    private fun loadShowAdditionalInfo(show: Show) {
        view.displayAdditionalInfoLoadingInProgress(true)
        val loadShowDataDisposable = showsRepository.getFullShowInfo(show.id)
                .compose(applySchedulersToSingle<Show>())
                .onErrorReturn { ignored -> Show.NOT_FOUND }
                .subscribe { showFullInfo ->
                    view.displayAdditionalInfoLoadingInProgress(false)
                    if (showFullInfo === Show.NOT_FOUND) {
                        view.displayAdditionalInfoNotLoaded(true)
                        return@subscribe
                    }

                    view.setShow(showFullInfo)
                    if (showFullInfo.seasons != null) {
                        view.displaySeasons(showFullInfo.seasons)
                    }
                    if (showFullInfo.cast != null) {
                        view.displayCastMembers(showFullInfo.cast)
                    }
                }

        addDisposable(loadShowDataDisposable)
    }

    private fun verifySavedShowStatus(show: Show) {
        val checkSavedShowDisposable = showsRepository.getSavedShow(show.id)
                .compose(applySchedulersToSingle<Show>())
                .subscribe({ savedShow -> view.displaySaveShowButton(false) }
                ) { ignored -> view.displaySaveShowButton(true) }

        addDisposable(checkSavedShowDisposable)
    }

    fun changeWatchingStatus(show: Show) {
        checkView()

        val checkStatusDisposable = showsRepository.getSavedShow(show.id)
                .compose(applySchedulersToSingle<Show>())
                .onErrorReturn { ignored -> Show.NOT_FOUND }
                .flatMap<Show> { showFromDb ->
                    if (showFromDb === Show.NOT_FOUND) {
                        showsRepository.saveShow(show)
                    } else {
                        showsRepository.removeShow(show.id)
                    }
                }
                .subscribe { showAfterChange ->
                    if (showAfterChange === Show.REMOVED) {
                        view.displayShowRemovedMessage()
                        view.displaySaveShowButton(true)
                    } else {
                        view.displaySavedShowMessage()
                        view.displaySaveShowButton(false)
                    }

                }

        addDisposable(checkStatusDisposable)
    }

    fun openEpisodes(show: Show, season: Season) {
        checkView()

        view.navigateToEpisodes(show, season.number)
    }
}
